tests/experimental/test_mcmc.py

```diff
@@ -87,23 +87,38 @@ def test_pCN_regression_warmup(target: cuqi.density.Density):
     sampler_new = cuqi.experimental.mcmc.pCNNew(target, scale=1)
     assert_true_if_warmup_is_equivalent(sampler_old, sampler_new)
 
+# ============ ULA ============
+
+@pytest.mark.parametrize("target", targets)
+def test_ULA_regression_sample(target: cuqi.density.Density):
+    """Test the ULA sampler regression."""
+    sampler_old = cuqi.sampler.ULA(target, scale=0.1)
+    sampler_new = cuqi.experimental.mcmc.ULANew(target, scale=0.1)
+    assert_true_if_sampling_is_equivalent(sampler_old, sampler_new)
+
+@pytest.mark.parametrize("target", targets)
+def test_MALA_regression_warmup(target: cuqi.density.Density):
+    """Test the ULA sampler regression."""
+    sampler_old = cuqi.sampler.ULA(target, scale=0.1)
+    sampler_new = cuqi.experimental.mcmc.ULANew(target, scale=0.1)
+    assert_true_if_warmup_is_equivalent(sampler_old, sampler_new)
+
 # ============ MALA ============
 
 @pytest.mark.parametrize("target", targets)
 def test_MALA_regression_sample(target: cuqi.density.Density):
-    """Test the pCN sampler regression."""
+    """Test the MALA sampler regression."""
     sampler_old = cuqi.sampler.MALA(target, scale=1)
     sampler_new = cuqi.experimental.mcmc.MALANew(target, scale=1)
     assert_true_if_sampling_is_equivalent(sampler_old, sampler_new)
 
 @pytest.mark.parametrize("target", targets)
 def test_MALA_regression_warmup(target: cuqi.density.Density):
-    """Test the pCN sampler regression."""
+    """Test the MALA sampler regression."""
     sampler_old = cuqi.sampler.MALA(target, scale=1)
     sampler_new = cuqi.experimental.mcmc.MALANew(target, scale=1)
     assert_true_if_warmup_is_equivalent(sampler_old, sampler_new)
 
-
 # ============ Checkpointing ============
 
 @pytest.mark.parametrize("sampler", [
```

Comment:
I suggest we add ULA to this set of tests. (and also MH and pCN). If MH and pCN fail, we can fix in another PR though.

Reply:
- Add ULA to checkpointing tests
- There seem 3 tiny bugs in the unit test function:
    -  `samples1` and `samples2` are both numpy arrays and thus they don't have `.samples` property (in the assert line)
    - the scale in `cuqi.experimental.mcmc.MALANew(cuqi.testproblem.Deconvolution1D().posterior, scale=1)` is too big and all the chains stay at the initial point, i.e., [1,1,...]
    - In defining `samples2`, we want to remove the first sample, instead of the last, so the code is changed to `samples2 = sampler_fresh.sample(100).get_samples().samples[...,1:]`.

Reply:
Neither MHNew or pCNNew passes this checkpointing test, and I'll add an issue to track this